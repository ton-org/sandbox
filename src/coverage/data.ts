import type {Step, TraceInfo} from "ton-assembly-dev-test/dist/trace"
import {Cell} from "@ton/core"

export type Coverage = {
    readonly code: Cell;
    readonly lines: readonly Line[];
};

export type Line = {
    readonly line: string
    readonly info: Covered | Uncovered | Skipped
}

export type Covered = {
    readonly $: "Covered"
    readonly hits: number
    readonly gasCosts: readonly number[]
}

export type Uncovered = {
    readonly $: "Uncovered"
}

export type Skipped = {
    readonly $: "Skipped"
}

export type InstructionStat = {
    readonly name: string
    readonly totalGas: number
    readonly totalHits: number
    readonly avgGas: number
}

export type CoverageSummary = {
    readonly totalLines: number
    readonly coveredLines: number
    readonly uncoveredLines: number
    readonly coveragePercentage: number
    readonly totalGas: number
    readonly totalHits: number
    readonly instructionStats: readonly InstructionStat[]
}

export const buildLineInfo = (trace: TraceInfo, asm: string): readonly Line[] => {
    const lines = asm.split("\n")

    const perLineSteps: Map<number, Step[]> = new Map()

    for (const step of trace.steps) {
        if (step.loc === undefined) continue
        const line = step.loc.line

        perLineSteps.set(line + 1, [...(perLineSteps.get(line + 1) ?? []), step])

        if (step.loc.otherLines.length > 0) {
            for (const otherLine of step.loc.otherLines) {
                perLineSteps.set(otherLine + 1, [...(perLineSteps.get(otherLine + 1) ?? []), step])
            }
        }
    }

    return lines.map((line, idx): Line => {
        const info = perLineSteps.get(idx + 1)
        if (info) {
            const gasInfo = info.map(step => normalizeGas(step.gasCost))

            return {
                line,
                info: {
                    $: "Covered",
                    hits: gasInfo.length,
                    gasCosts: gasInfo,
                },
            }
        }

        if (!isExecutableLine(line)) {
            return {
                line,
                info: {
                    $: "Skipped",
                },
            }
        }

        return {
            line,
            info: {
                $: "Uncovered",
            },
        }
    })
}

const normalizeGas = (gas: number): number => {
    if (gas > 10000) {
        return 26
    }
    return gas
}

export const isExecutableLine = (line: string): boolean => {
    const trimmed = line.trim()
    return (
        !trimmed.includes("=>") && // dictionary
        trimmed !== "}" && // close braces
        trimmed !== "]" && // close bracket
        trimmed.length > 0
    )
}

export const generateCoverageSummary = (
    coverage: Coverage,
): CoverageSummary => {
    const lines = coverage.lines;
    const totalExecutableLines = lines.filter(line => isExecutableLine(line.line)).length

    const coveredLines = lines.filter(
        line => isExecutableLine(line.line) && line.info.$ === "Covered",
    ).length
    const uncoveredLines = totalExecutableLines - coveredLines
    const coveragePercentage = totalExecutableLines === 0 ? 0 : (coveredLines / totalExecutableLines) * 100

    let totalGas = 0
    let totalHits = 0

    const instructionMap: Map<string, { readonly totalGas: number; readonly hits: number }> =
        new Map()

    for (const line of lines) {
        if (line.info.$ !== "Covered") continue

        const lineGas = line.info.gasCosts.reduce((sum, gas) => sum + gas, 0)
        totalGas += lineGas
        totalHits += line.info.hits
        const trimmedLine = line.line.trim()
        const instructionName = trimmedLine.split(/\s+/)[0]
        if (instructionName !== undefined) {
            const current = instructionMap.get(instructionName) ?? {totalGas: 0, hits: 0}
            instructionMap.set(instructionName, {
                totalGas: current.totalGas + lineGas,
                hits: current.hits + line.info.hits,
            })
        }
    }

    const instructionStats: InstructionStat[] = [...instructionMap.entries()]
        .map(([name, stats]) => ({
            name,
            totalGas: stats.totalGas,
            totalHits: stats.hits,
            avgGas: stats.hits === 0 ? 0 : Math.round((stats.totalGas / stats.hits) * 100) / 100,
        }))
        .sort((a, b) => b.totalGas - a.totalGas)

    return {
        totalLines: totalExecutableLines,
        coveredLines,
        uncoveredLines,
        coveragePercentage,
        totalGas,
        totalHits,
        instructionStats,
    }
}

export const mergeCoverages = (...coverages: readonly Coverage[]): Coverage => {
    if (coverages.length === 0) {
        return {
            code: new Cell(),
            lines: [],
        };
    }

    let allLines: readonly Line[] = coverages[0]?.lines ?? [];
    for (const coverage of coverages.slice(1)) {
        allLines = mergeTwoLines(allLines, coverage.lines);
    }
    return {
        code: coverages[0]?.code ?? new Cell(),
        lines: allLines,
    };
};

export const mergeTwoLines = (
    first: readonly Line[],
    second: readonly Line[],
): readonly Line[] => {
    if (first.length !== second.length) return first;

    const result: Line[] = [...first];

    for (const [index, line] of second.entries()) {
        const prev = result[index];
        if (!prev) continue;

        if (prev.info.$ === "Uncovered" && line.info.$ === "Uncovered") {
            // nothing changes
            continue;
        }

        if (prev.info.$ === "Skipped" && line.info.$ === "Skipped") {
            // nothing changes
            continue;
        }

        if (prev.info.$ === "Uncovered" && line.info.$ === "Covered") {
            // replace it with new data
            result[index] = line;
        }

        if (prev.info.$ === "Covered" && line.info.$ === "Uncovered") {
            // nothing changes
            continue;
        }

        if (prev.info.$ === "Covered" && line.info.$ === "Covered") {
            result[index] = {
                ...prev,
                info: {
                    ...prev.info,
                    hits: prev.info.hits + line.info.hits,
                    gasCosts: [...prev.info.gasCosts, ...line.info.gasCosts],
                },
            };
        }
    }

    return result;
};

export const coverageToJson = (coverage: Coverage): string => {
    const lines = coverage.lines;
    return JSON.stringify({
        code: coverage.code.toBoc().toString("hex"),
        lines: lines.map((line, index) => {
            if (line.info.$ === "Covered") {
                return {
                    lineNumber: index,
                    line: line.line,
                    info: {
                        ...line.info,
                    },
                };
            }
            return {
                lineNumber: index,
                ...line,
            };
        }),
    });
};

export const coverageFromJson = (string: string): Coverage => {
    type CoverageJson = {
        readonly code: string;
        readonly lines: readonly Line[];
    };

    const data = JSON.parse(string) as CoverageJson;
    return {
        code: Cell.fromHex(data.code),
        lines: data.lines
    };
};
