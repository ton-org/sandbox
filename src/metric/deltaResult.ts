import {
    AddressFriendly,
    ContractMethodName,
    ContractName,
    Metric,
    OpCode,
    SnapshotMetric,
    sortByCreatedAt,
} from './collectMetric';

export type KindDelta = 'undefined' | 'init' | 'same' | 'increase' | 'decrease';

export type PathDelta = string;

export type ItemDelta = {
    kind: KindDelta;
    path: PathDelta;
    before: number;
    after: number;
};

export type ListDelta = Record<PathDelta, ItemDelta>;

export type DeltaMetric = {
    kind: KindDelta;
    value: string;
};

export const undefinedDeltaMetric = () =>
    ({
        kind: 'undefined',
        value: '~',
    }) as DeltaMetric;

export type ColorDelta = (metric: DeltaMetric) => string;

export type DeltaMetrics = Record<string, DeltaMetric>;

export type MethodDelta = Record<ContractMethodName | OpCode, DeltaMetrics>;

export type ContractDelta = Record<ContractName | AddressFriendly, MethodDelta>;

export type DeltaResult = {
    label: string;
    createdAt: Date;
    result: ContractDelta;
};

export type DeltaRow = [contract: string, method: string, ...values: DeltaMetric[]];

export type FlatDeltaResult = {
    header: string[];
    rows: DeltaRow[];
};

export function toFlatDeltaResult(deltas: DeltaResult[]): { header: string[]; rows: DeltaRow[] } {
    const out: FlatDeltaResult = {
        header: ['Contract', 'Method'],
        rows: [],
    };
    const contractSet = new Set<string>();
    const methodMap = new Map<string, Set<string>>();
    const metricSet = new Set<string>();

    for (const delta of deltas) {
        for (const [contract, methods] of Object.entries(delta.result)) {
            contractSet.add(contract);
            if (!methodMap.has(contract)) {
                methodMap.set(contract, new Set());
            }

            for (const [method, metrics] of Object.entries(methods)) {
                methodMap.get(contract)!.add(method);

                for (const metric of Object.keys(metrics)) {
                    metricSet.add(metric);
                }
            }
        }
    }
    const metricNames = Array.from(metricSet.values());
    out.header.push(...Array.from({ length: deltas.length }, () => metricNames).flat());

    for (const contract of contractSet) {
        const methods = methodMap.get(contract)!;
        for (const method of methods) {
            const metrics: DeltaMetric[] = [];
            for (const delta of deltas) {
                for (const metric of metricNames) {
                    metrics.push(delta.result?.[contract]?.[method]?.[metric] ?? undefinedDeltaMetric());
                }
            }
            out.rows.push([contract, method, ...metrics]);
        }
    }
    return out;
}

const aggregateTarget = ['cells', 'bits', 'gasUsed'];
const statusTarget = ['success', 'exitCode', 'resultCode'];
const deltaTarget = [
    ...aggregateTarget,
    ...statusTarget,
    'vmSteps',
    'totalActions',
    'skippedActions',
    'totalActionFees',
    'data',
    'code',
    'state',
    'message',
    'in',
    'out',
    'execute',
    'compute',
    'action',
];

/**
 * Recursively collects the sum of all numeric fields named `needle` within an arbitrary data structure
 */
function sumDeep(source: unknown, needle: string): number {
    if (source === null || typeof source !== 'object') {
        return 0;
    }

    if (Array.isArray(source)) {
        return source.map((item) => sumDeep(item, needle)).reduce((total, current) => total + current, 0);
    }

    let total = 0;
    for (const [key, value] of Object.entries(source as Record<string, unknown>)) {
        if (key === needle && typeof value === 'number') {
            total += value;
            continue;
        }
        total += sumDeep(value, needle);
    }

    return total;
}

export function aggregatedCompareMetric(before: Metric, after: Metric, basePath: string[] = []): ListDelta {
    const out: ListDelta = {};
    const keys = new Set<string>([...Object.keys(before ?? {}), ...Object.keys(after ?? {})]);
    keys.forEach((key) => {
        if (!deltaTarget.includes(key)) return;

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const prev = (before as any)[key];
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const next = (after as any)[key];
        const path = [...basePath, key];

        if (prev === undefined || next === undefined) return;

        if (typeof prev === 'object' && typeof next === 'object') {
            Object.assign(out, aggregatedCompareMetric(prev, next, path));
            aggregateTarget.forEach((aggKey) => {
                const beforeSum = sumDeep(prev, aggKey);
                const afterSum = sumDeep(next, aggKey);
                const aggPath = [...path, aggKey].join('.');
                out[aggPath] = {
                    kind: beforeSum > afterSum ? 'decrease' : beforeSum < afterSum ? 'increase' : 'same',
                    path: aggPath,
                    before: beforeSum,
                    after: afterSum,
                };
            });
            return;
        }

        if (typeof prev === 'number' && typeof next === 'number') {
            const fullPath = path.join('.');
            out[fullPath] = {
                kind: prev > next ? 'decrease' : prev < next ? 'increase' : 'same',
                path: fullPath,
                before: prev,
                after: next,
            };
        }
    });

    if (basePath.length === 0) {
        aggregateTarget.forEach((aggKey) => {
            const beforeSum = sumDeep(before, aggKey);
            const afterSum = sumDeep(after, aggKey);
            const rootPath = `root.${aggKey}`;
            out[rootPath] = {
                kind: beforeSum > afterSum ? 'decrease' : beforeSum < afterSum ? 'increase' : 'same',
                path: rootPath,
                before: beforeSum,
                after: afterSum,
            };
        });
    }
    return out;
}

function prepareItemDelta(item?: ItemDelta, calcDelta = true): DeltaMetric {
    const out: DeltaMetric = undefinedDeltaMetric();
    if (!item) {
        return out;
    }
    out.kind = calcDelta ? item.kind : 'init';
    out.value = item.after.toString();
    if (calcDelta) {
        let change = item.kind === 'increase' ? ' +' : ' ';
        if (item.kind === 'same') {
            change += 'same';
        } else if (item.before === 0) {
            change += '100.00%';
        } else {
            change += (((item.after - item.before) / item.before) * 100).toFixed(2) + '%';
        }
        out.value += change;
    }
    return out;
}

export function prepareDelta(pair: { after: SnapshotMetric; before?: SnapshotMetric }): DeltaResult {
    const result: ContractDelta = {};
    const out: DeltaResult = {
        label: pair.after.label,
        createdAt: pair.after.createdAt,
        result,
    };

    const beforeMap = new Map<string, Map<string, Metric>>();
    if (pair.before) {
        for (const b of pair.before.items) {
            const contractKey = b.contractName || b.address;
            const methodKey = b.methodName || b.opCode;
            if (!beforeMap.has(contractKey)) {
                beforeMap.set(contractKey, new Map());
            }
            beforeMap.get(contractKey)!.set(methodKey, b);
        }
    }

    for (const a of pair.after.items) {
        const contractKey = a.contractName || a.address;
        const methodKey = a.methodName || a.opCode;

        const b = beforeMap.get(contractKey)?.get(methodKey); // может быть undefined
        const calcDelta = !!b;

        if (!result[contractKey]) {
            result[contractKey] = {};
        }

        const item: Record<string, DeltaMetric> = {};
        result[contractKey][methodKey] = item;

        const delta = aggregatedCompareMetric(b || a, a);
        item.gasUsed = prepareItemDelta(delta['execute.compute.gasUsed'], calcDelta);
        item.cells = prepareItemDelta(delta['state.cells'], calcDelta);
        item.bits = prepareItemDelta(delta['state.bits'], calcDelta);
    }

    return out;
}

export function makeGasReport(list: SnapshotMetric[]): Array<DeltaResult> {
    list = (list || []).sort(sortByCreatedAt());
    const out: DeltaResult[] = [];
    if (list.length === 0) return out;
    for (let i = 0; i < list.length; i++) {
        const after = list[i];
        const before = i > 0 ? list[i - 1] : undefined;
        out.push(prepareDelta({ after, before }));
    }
    return out;
}
