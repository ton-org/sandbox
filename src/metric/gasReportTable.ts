import { table, TableUserConfig, ColumnUserConfig, SpanningCellConfig } from 'table';
import { ColorDelta, DeltaResult, toFlatDeltaResult } from './deltaResult';
import { sortByCreatedAt } from './collectMetric';

const gapCellWidth = 1;
const maxCellWidth = 25;
const border = {
    topBody: '─',
    topJoin: '┬',
    topLeft: '┌',
    topRight: '┐',
    bottomBody: '─',
    bottomJoin: '┴',
    bottomLeft: '└',
    bottomRight: '┘',
    bodyLeft: '│',
    bodyRight: '│',
    bodyJoin: '│',
    joinBody: '─',
    joinLeft: '├',
    joinRight: '┤',
    joinJoin: '┼',
};

function wrap(value: string, max: number) {
    if (value.length > max) {
        return `${value.slice(0, max - 3)}...`;
    }
    return value;
}

function prepareResult(list: DeltaResult[], color?: ColorDelta) {
    list = list.sort(sortByCreatedAt(true));

    const flat = toFlatDeltaResult(list);
    const rows: string[][] = [];
    const widthCols: number[] = [];

    for (const [contract, method, ...values] of flat.rows) {
        const row: string[] = [wrap(contract, maxCellWidth), wrap(method, maxCellWidth)];
        widthCols[0] = Math.max(widthCols[0] ?? 0, row[0].length);
        widthCols[1] = Math.max(widthCols[1] ?? 0, row[1].length);
        values.forEach((metric, idx) => {
            const value = color ? color(metric) : metric.value;
            row.push(value);
            const colIdx = idx + 2;
            widthCols[colIdx] = Math.max(widthCols[colIdx] ?? 0, metric.value.length);
        });
        rows.push(row);
    }
    const headers = flat.header;
    for (let i = 0; i < headers.length; i++) {
        widthCols[i] = Math.max(widthCols[i] ?? 0, headers[i].length + gapCellWidth);
    }
    const groupIndex: Record<string, { index: number; size: number }> = {};
    let current = 0;
    while (current < rows.length) {
        const contract = rows[current][0];
        let count = 1;
        while (rows[current + count]?.[0] === contract) {
            count++;
        }
        groupIndex[contract] = { index: current, size: count };
        current += count;
    }

    return {
        labels: list.map((s) => s.label),
        headers,
        widthCols,
        rows,
        groupIndex,
    };
}

export function gasReportTable(list: DeltaResult[], color?: ColorDelta) {
    const result = prepareResult(list, color);
    if (result.rows.length < 1) {
        return 'No data available';
    }
    const columns: ColumnUserConfig[] = [];
    for (let i = 0; i < result.headers.length; i++) {
        columns.push({ alignment: 'center', verticalAlignment: 'middle', width: result.widthCols[i] });
    }
    const spanningCells: SpanningCellConfig[] = [
        { col: 0, row: 0, rowSpan: 2, verticalAlignment: 'middle' }, // Contract title
        { col: 1, row: 0, rowSpan: 2, verticalAlignment: 'middle' }, // Method title
    ];
    for (const group of Object.values(result.groupIndex)) {
        // rowSpan for Contract name
        spanningCells.push({ col: 0, row: group.index + 2, rowSpan: group.size, verticalAlignment: 'middle' });
    }
    const data: string[][] = [[], ['', '']];
    data[0].push(...result.headers.slice(0, 2));
    data[1].push(...result.headers.slice(2));
    const metricCount = (result.headers.length - 2) / list.length;
    let labelTitleOffset = 2;
    for (const label of result.labels) {
        spanningCells.push({
            col: labelTitleOffset,
            row: 0,
            colSpan: metricCount,
            verticalAlignment: 'middle',
        });
        labelTitleOffset += metricCount;
        data[0].push(...[label, ...Array.from({ length: metricCount - 1 }, () => '')]);
    }
    data.push(...result.rows);
    const config: TableUserConfig = {
        columns,
        spanningCells,
        border,
    };
    return table(data, config);
}
