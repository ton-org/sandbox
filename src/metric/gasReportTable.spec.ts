import { gasReportTable } from './gasReportTable';
import { oneFirst, oneSecond, zeroFirst } from './fixtures/data.fixture';
import { DeltaResult, makeGasReport } from './deltaResult';
import { sampleList } from './fixtures/snapshot.fixture';
import complexList from './fixtures/complex.fixture';

describe('gasReportTable', () => {
    it.each([
        { list: new Array<DeltaResult>() },
        {
            list: [
                {
                    label: 'empty',
                    createdAt: new Date('2009-01-03T00:00:00Z'),
                    result: {},
                } as DeltaResult,
            ],
        },
        {
            list: [
                {
                    label: 'empty',
                    createdAt: new Date('2009-01-03T00:00:00Z'),
                    result: { contractName: {} },
                },
            ],
        },
    ])('empty data', (data) => {
        const actual = gasReportTable(data.list);
        expect(actual).toEqual('No data available');
    });

    it('table single result', () => {
        const actual = gasReportTable(makeGasReport([oneFirst]));
        expect(actual).toMatchSnapshot();
    });

    it('table delta result same', () => {
        const actual = gasReportTable(makeGasReport([oneSecond, oneFirst]));
        expect(actual).toMatchSnapshot();
    });

    it('table delta result increase', () => {
        const actual = gasReportTable(makeGasReport([oneSecond, zeroFirst]));
        expect(actual).toMatchSnapshot();
    });

    it('table sample', () => {
        const actual = gasReportTable(makeGasReport(sampleList));
        expect(actual).toMatchSnapshot();
    });

    it('table complex', () => {
        const actual = gasReportTable(makeGasReport(complexList));
        expect(actual).toMatchSnapshot();
    });
});
