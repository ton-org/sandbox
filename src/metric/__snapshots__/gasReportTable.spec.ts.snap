// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gasReportTable table complex 1`] = `
"┌───────────────┬───────────────────────┬──────────────────────────────────────┬───────────────────────────┐
│               │                       │               current                │            v1             │
│   Contract    │        Method         ├──────────────┬─────────┬─────────────┼──────────┬────────┬───────┤
│               │                       │   gasUsed    │  cells  │    bits     │ gasUsed  │ cells  │ bits  │
├───────────────┼───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      sendDeploy       │  1937 same   │ 36 same │  7650 same  │   1937   │   36   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │         send          │   489 same   │ 36 same │  7650 same  │   489    │   36   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │     sendDeployNft     │  1937 same   │ 36 same │  7650 same  │   1937   │   36   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │          0x1          │  7251 same   │ 36 same │  7650 same  │   7251   │   36   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x800637ac       │ 3208 -0.31%  │ 36 same │ 7459 +1.47% │   3218   │   36   │ 7351  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │ sendGetRoyaltyParams  │  1937 same   │ 37 same │ 7758 +1.41% │   1937   │   37   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x693d3950       │ 3126 +10.42% │ 37 same │ 7758 +1.41% │   2831   │   37   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0xa8cb00ad       │   309 same   │ 37 same │ 7758 +1.41% │   309    │   37   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x800f8178       │     3208     │   37    │    7758     │    ~     │   ~    │   ~   │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │     notSupported      │ 849 +14.57%  │ 37 same │ 7758 +1.41% │   741    │   37   │ 7650  │
│ NFTCollection ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x8008751c       │  3218 same   │ 36 same │  7650 same  │   3218   │   36   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │  sendBatchDeployNFT   │  1937 same   │ 36 same │  7650 same  │   1937   │   36   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │          0x2          │ 727481 same  │ 36 same │  7650 same  │  727481  │   36   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x801a18cb       │     3208     │   37    │    7758     │    ~     │   ~    │   ~   │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x800299f8       │     3208     │   37    │    7758     │    ~     │   ~    │   ~   │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │    sendChangeOwner    │  1937 same   │ 37 same │ 7758 +1.41% │   1937   │   37   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │          0x3          │ 2582 +5.47%  │ 37 same │ 7758 +1.41% │   2448   │   37   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x8017bad0       │      ~       │    ~    │      ~      │   3218   │   37   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x80105974       │      ~       │    ~    │      ~      │   3218   │   37   │ 7650  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x8002d39c       │      ~       │    ~    │      ~      │   3218   │   37   │ 7650  │
├───────────────┼───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │ sendTransferOwnership │  1937 same   │ 15 same │  3782 same  │   1937   │   15   │ 3782  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x5fcc3d14       │  6850 same   │ 15 same │  3782 same  │   6850   │   15   │ 3782  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0xd53276db       │   309 same   │ 15 same │  3782 same  │   309    │   15   │ 3782  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │       0x5138d91       │    0 same    │ 16 same │ 3983 -1.63% │    0     │   16   │ 4049  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │     notSupported      │   309 same   │ 16 same │ 3983 -1.63% │   309    │   16   │ 4049  │
│    NFTItem    ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │   sendGetStaticData   │  1937 same   │ 15 same │  3782 same  │   1937   │   15   │ 3782  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x2fcb26a2       │  4226 same   │ 15 same │  3782 same  │   4226   │   15   │ 3782  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x8b771735       │   309 same   │ 15 same │  3782 same  │   309    │   15   │ 3782  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      sendDeploy       │  1937 same   │ 15 same │  3782 same  │   1937   │   15   │ 3782  │
│               ├───────────────────────┼──────────────┼─────────┼─────────────┼──────────┼────────┼───────┤
│               │      0x8008751c       │  3218 same   │ 15 same │  3782 same  │   3218   │   15   │ 3782  │
└───────────────┴───────────────────────┴──────────────┴─────────┴─────────────┴──────────┴────────┴───────┘
"
`;

exports[`gasReportTable table delta result increase 1`] = `
"┌───────────────────────────┬─────────┬──────────────────────────────────────┬───────────────────────────┐
│                           │         │              one second              │        zero first         │
│         Contract          │ Method  ├────────────┬────────────┬────────────┼──────────┬────────┬───────┤
│                           │         │  gasUsed   │   cells    │    bits    │ gasUsed  │ cells  │ bits  │
├───────────────────────────┼─────────┼────────────┼────────────┼────────────┼──────────┼────────┼───────┤
│ EQBGhqLAZseEqRXz4ByFPT... │   0x0   │ 1 +100.00% │ 2 +100.00% │ 2 +100.00% │    0     │   0    │   0   │
└───────────────────────────┴─────────┴────────────┴────────────┴────────────┴──────────┴────────┴───────┘
"
`;

exports[`gasReportTable table delta result same 1`] = `
"┌───────────────────────────┬─────────┬────────────────────────────┬───────────────────────────┐
│                           │         │         one second         │         one first         │
│         Contract          │ Method  ├──────────┬────────┬────────┼──────────┬────────┬───────┤
│                           │         │ gasUsed  │ cells  │  bits  │ gasUsed  │ cells  │ bits  │
├───────────────────────────┼─────────┼──────────┼────────┼────────┼──────────┼────────┼───────┤
│ EQBGhqLAZseEqRXz4ByFPT... │   0x0   │  1 same  │ 2 same │ 2 same │    1     │   2    │   2   │
└───────────────────────────┴─────────┴──────────┴────────┴────────┴──────────┴────────┴───────┘
"
`;

exports[`gasReportTable table sample 1`] = `
"┌─────────────┬────────────────┬────────────────────────────────────┬───────────────────────────┐
│             │                │              current               │           first           │
│  Contract   │     Method     ├─────────────┬────────┬─────────────┼──────────┬────────┬───────┤
│             │                │   gasUsed   │ cells  │    bits     │ gasUsed  │ cells  │ bits  │
├─────────────┼────────────────┼─────────────┼────────┼─────────────┼──────────┼────────┼───────┤
│             │   sendDeploy   │  1937 same  │ 9 same │ 1023 -4.48% │   1937   │   9    │ 1071  │
│             ├────────────────┼─────────────┼────────┼─────────────┼──────────┼────────┼───────┤
│             │      send      │ 589 -8.11%  │ 9 same │ 1023 -4.48% │   641    │   9    │ 1071  │
│             ├────────────────┼─────────────┼────────┼─────────────┼──────────┼────────┼───────┤
│             │ sendIncrAction │  1937 same  │ 9 same │ 1023 -4.48% │   1937   │   9    │ 1071  │
│ SampleBench ├────────────────┼─────────────┼────────┼─────────────┼──────────┼────────┼───────┤
│             │      Incr      │ 2743 -1.58% │ 9 same │ 1023 -4.48% │   2787   │   9    │ 1071  │
│             ├────────────────┼─────────────┼────────┼─────────────┼──────────┼────────┼───────┤
│             │ sendDecrAction │  1937 same  │ 9 same │ 1023 -4.48% │   1937   │   9    │ 1071  │
│             ├────────────────┼─────────────┼────────┼─────────────┼──────────┼────────┼───────┤
│             │      Decr      │ 2820 -2.25% │ 9 same │ 1023 -4.48% │   2885   │   9    │ 1071  │
└─────────────┴────────────────┴─────────────┴────────┴─────────────┴──────────┴────────┴───────┘
"
`;

exports[`gasReportTable table single result 1`] = `
"┌───────────────────────────┬─────────┬───────────────────────────┐
│                           │         │         one first         │
│         Contract          │ Method  ├──────────┬────────┬───────┤
│                           │         │ gasUsed  │ cells  │ bits  │
├───────────────────────────┼─────────┼──────────┼────────┼───────┤
│ EQBGhqLAZseEqRXz4ByFPT... │   0x0   │    1     │   2    │   2   │
└───────────────────────────┴─────────┴──────────┴────────┴───────┘
"
`;
